### ISSUE TITLE
Finish refactoring from UBKG API

### ISSUE THREAD

**Issue by shirey** [x-atlas-consortia/hs-ontology-api#14]
Finish the work started by Sam to separate all components and code specific to this HuBMAP/SenNet specific API/service out of the central [UBKG API](https://github.com/x-atlas-consortia/ubkg-api) code.

@ChuckKollar Please fill in any details that you noted when you met with Sam about this before he left.

This issue needs to be finished before #6 #7 and #8 can be completed.

**Comment by ChuckKollar** [comment]
Notes from Sam on Jul 14, 2023

Joe mentioned in standup that sometime that you access it returns a 500 and other times it works.
Not sure.

hs-ontology.api
Pulled in queries from neo4j manager (common) to neo4j_logic.py (hg-) valuset organs and datasets
and removed them from the common repo.

The common repo (ubkg-api) is used for whatever other applications want to use it.
Outside of our team there is some other team that is using the common code. Alan would know about that.

What is the overall design?
Sam just got a task to take out the "openapi stuff", and that's the only design that he knows of.
There is a specific repo and there is a common repo used by all.
The openapi files are managed by hand.
Main (hg-) and app (common) is the place where the blueprints are pulled together.
The Neo4J connection is created in neo4j_manager.py in the common is instantiated there and attached to the flask app.
The neo4j_manager.py has the life cycle code for the neo4j as well as the "common query" methods,
there are also some common things like validation there.
In the specific repo the Neo4J_Manager is instantiated and attached to the app.
All specific Neo4J queries are in functions in the specific repo and they take a parameter which is the
neo4j object. The routs get this neo4j object from the current_app ane pass it to these functions which
return the result of a neo4j call.

**Comment by AlanSimmons** [comment]
The original request was to separate the endpoint code in the UBKG API into application context-specific APIs that could also use common endpoint code. All APIs would work with the same instance of a UBKG neo4j instance. 

The UBKG would be like a base class. The application context-specific APIs would be like superclasses--i.e., they would inherit endpoints from the UBKG, but have endpoints that are specific to the context. The "superclasses" would contain both manager code and routes.

The current application contexts are:
HuBMAP/SenNet
Data Distillery

This repo (hs-ontology-api) corresponds to the API specific to the HuBMAP/SenNet application context. 

The hs-ontology-api should support endpoints that are specific to HuBMAP and SenNet--e.g., datasets; organ; valueset; etc. The UBKG API (base class) should support the endpoints that were in the original ontology api, along with any that we develop that we think will be useful to all contexts. 

Right now, code for all endpoints is still in the UBKG API. The hs-ontology-api only contains routes. 

"Finishing the refactoring" entails moving the endpoint code from the UBKG API manager to the hs-ontology-api manager.

**Comment by ChuckKollar** [comment]
Currently there is a branch in UBKG-api that Sam was working on and that I debug to get it working while Sam watched. When you instantiate UBKG-api it actually returns a Flask app, which I have never seen done in a "library", and it causes some confusion to someone writing  the Flask app using this library (e.g., you give up some control over the app to the library). The other confusion is that the flask app created by UBKG-api places the Neo4J connection on the 'current_app' so it can only be accessed from the endpoints and passed to the functions that do the Neo4J queries by the apps that us it like hs-ontoloty-api. However, since the Neo4J access in UBKG-app is through methods in the Neo4j object it access the Neo4J instance directly from the object that it is in. This leads to confusion to anyone looking at the "library" to "extend it" because they cannot use the neo4j access methods in the same way that the library does. This cause another bit of the problem that I had to work around while debugging this. There is also an overlapping namespace issue that I needed to riddle out and fix.

While the current code "appears" to work. I would consider making it much less confusing by having the UBKG-api create only blueprints and have one function that wires these blueprints to the Flask app using them (e.g., hs-ontoloty-api) so that the user has full control over the Flask app. I would also split out the Neo4j access methods into functions that are passed the Neo4j connection (like you need to do in the project using this sub project).  The other thing that I would do is to get rid of the Objects that were created by the OpenAPI code that built the App from the OpenAPI yaml file. This is a good pattern to follow when the micro-service is being autogenerated (you get requests/responses that you access like a class but you don't maintain it as it is autogenerated), but when you are not autogenerating things and keeping the OpenAPI in sync with the code automatically, it is much simpler to just return a json object directly. The down side is that the json objects are more error prone, but the classes that came for free to help you avoid errors are no longer free and no one is going to create them moving forward because they are too much work to maintain by hand.

Finally, it was never explained to me why the original work that I did was not extended as I had suggested that it could be. It was really simple to use. Alan figured out how to use it and add endpoints without my help. You simply created a method in the Neo4j manager that had the same name as the flask endpoint, put your neo4j code there and returned a class (response object defined by the OpenAPI file) that would be converted to json automatically. It kept the OpenAPI spec in sync with the code and extending it was extremely simple. I consider this original implementation to be a vast improvement on what was created to replace it.

**Comment by ChuckKollar** [comment]
I discussed the changes that I had proposed with Bill, and he agreed that I should proceed forward on them. I have two pull requests and when the code is merged into dev-integrate, I will test them.

**Comment by AlanSimmons** [comment]
# Results of testing using test scripts: Errors

@ChuckKollar 

Reviewed in dev instance ( https://ontology-api.dev.hubmapconsortium.org/), using calls mentioned in the following test scripts: 

1.  [hs-ontology-api test script](https://github.com/x-atlas-consortia/hs-ontology-api/blob/lifting_state_for_clearset_all_in_model/src/test_api.sh).
2. [ubkg test script](https://github.com/x-atlas-consortia/ubkg-api/blob/2-refactor-application-endpoints-with-Bill-suggestions/src/ubkg_api/test_api.sh)

Issues:

1. Both test scripts refer to a bearer token. There is no need for a bearer token; both APIs are open. I can get results by putting the URLs in a browser window.
2. No URLs in the UBKG script should include application_context as a parameter. Because of the refactoring, the UBKG should not include any endpoints that depend on application context. I can get the same results in calls that are in UBKG regardless of whether I supply an application_context parameter value.
3. The example for valuesets in hs-ontology-api does not return anything, because parent_code=C000530 in HuBMAP does not correspond to a valueset. This URL, on the other hand, will return something: https://ontology-api.dev.hubmapconsortium.org/valueset?child_sabs=HGNC&child_sabs=OBI&parent_sab=HUBMAP&parent_code=C001000

In addition, @yuanzhou mentioned the following:

4.[Bill Shirey](https://hubmapconsortium.slack.com/team/UE4AZ00SD)
 suggested that we probably should put the script outside the src directory too, say under the project root directory
when I tried the test script, I had to edit the file so it points to either DEV/PROD or my local UBKG-API URL. Can we take the target URL as an argument/option in the command line to avoid editing the script?
5. here are ingest-api URLs and search-api URLs in the scripts, but we don't use them at all in UBKG-API tests

**Comment by AlanSimmons** [comment]
The UBKG API still contains code that should be in hs-ontology-api.

Breadcrumbs[ubkg-api](https://github.com/x-atlas-consortia/ubkg-api/tree/2-refactor-application-endpoints-with-Bill-suggestions)/[src](https://github.com/x-atlas-consortia/ubkg-api/tree/2-refactor-application-endpoints-with-Bill-suggestions/src)/[ubkg_api](https://github.com/x-atlas-consortia/ubkg-api/tree/2-refactor-application-endpoints-with-Bill-suggestions/src/ubkg_api)/[common_routes](https://github.com/x-atlas-consortia/ubkg-api/tree/2-refactor-application-endpoints-with-Bill-suggestions/src/ubkg_api/common_routes)
/common_neo4j_logic.py

Everything below the comment should be in hs-ontology-api:
```
# -----------------------------
# HUBMAP/SENNET ENDPOINTS
```

**Comment by AlanSimmons** [comment]
Creating a new ticket.

**Comment by AlanSimmons** [comment]
Per @yuanzhou, the refactored ubkg-api/hs-ontology-api was released today.
